cumulusci/tasks/connectedapp.py

```diff
@@ -0,0 +1,174 @@
+                )
+            self.options["email"] = github.email
+
+        # Default to sfdx defaultdevhubusername
+        if "username" not in self.options:
+            self._set_default_username()
+        self.options["command"] += " -u {}".format(self.options.get("username"))
+
+        self.options["connect"] = process_bool_arg(self.options.get("connect", False))
+        self.options["overwrite"] = process_bool_arg(
+            self.options.get("overwrite", False)
+        )
+
+    def _set_default_username(self):
+        self.logger.info("Getting username for the default devhub from sfdx")
+        self._run_command(
+            command="{} force:config:get defaultdevhubusername --json".format(SFDX_CLI),
+            env=self._get_env(),
+            output_handler=self._process_devhub_output,
+        )
+
+    def _process_json_output(self, line):
+        try:
+            data = json.loads(line)
+            return data
+        except Exception:
+            self.logger.error("Failed to parse json from line: {}".format(line))
+            raise
+
+    def _process_devhub_output(self, line):
+        data = self._process_json_output(line)
+        if "value" not in data["result"][0]:
+            raise TaskOptionsError(
+                "No sfdx config found for defaultdevhubusername.  Please use the sfdx force:config:set to set the defaultdevhubusername and run again"
+            )
+        self.options["username"] = data["result"][0]["value"]
+
+    def _generate_id_and_secret(self):
+        self.client_id = random_alphanumeric_underscore(85)
+        self.client_secret = random_alphanumeric_underscore(32)
+
+    def _build_package(self):
+        connected_app_path = os.path.join(self.tempdir, "connectedApps")
+        os.mkdir(connected_app_path)
+        self._generate_id_and_secret()
+        with open(
+            os.path.join(connected_app_path, self.options["label"] + ".connectedApp"),
+            "w",
+        ) as f:
+            f.write(
+                CONNECTED_APP.format(
+                    **{
+                        "label": self.options["label"],
+                        "email": self.options["email"],
+                        "client_id": self.client_id,
+                        "client_secret": self.client_secret,
+                    }
+                )
+            )
+        with open(os.path.join(self.tempdir, "package.xml"), "w") as f:
+            f.write(PACKAGE_XML)
+
+    def _connect_service(self):
+        if not self.options["overwrite"]:
+            try:
+                connected_app = self.project_config.keychain.get_service(
+                    "connected_app"
+                )
+                raise TaskOptionsError(
+                    "The CumulusCI keychain already contains a connected_app service.  Set the 'overwrite' option to True to overwrite the existing service"
```

Comment:
so in this case a connected app still was created and deployed to the devhub, it just doesn't overwrite the cci service? i guess that works.

Reply:
I should probably change that to validate that it can write and not deploy if not